### DISCUSSION TITLE
Install Guide for Turnkey VMs in Proxmox VE Helper-Scripts

### DISCUSSION THREAD

**Discussion by havardthom** [community-scripts/ProxmoxVE#144] [Guides]
This guide will explain how to fully install a Turnkey VM using [Proxmox VE Helper-Scripts](https://community-scripts.github.io/ProxmoxVE). These VMs comes with a Turnkey ISO image bootdisk and requires some extra installation steps to get the VM fully up and running.

- Start by installing a Turnkey VM using [Proxmox VE Helper-Scripts](https://community-scripts.github.io/ProxmoxVE/). In this example I'll be installing a Turnkey Nextcloud VM: 
   `bash -c "$(wget -qLO - https://github.com/community-scripts/ProxmoxVE/raw/main/vm/nextcloud-vm.sh)"`
     - The VM is created with two disks by default:
          - `scsi0` - OS disk
          - `scsi1` - Turnkey ISO image disk
    
![hardware](https://github.com/user-attachments/assets/aba88f75-a650-48de-87e1-b38feb390af0)

(Expand sections for screenshots)

<details>
<summary>  <b>(Optional)</b> Before booting, resize the OS disk (scsi0) to your liking, since it will be partitioned on first boot</summary>

![resize-disk](https://github.com/user-attachments/assets/5b9c7df2-6dd2-4676-9185-dabb461a5bdf)

</details>

<details>
<summary>Start the VM (boots from Turnkey ISO image disk by default)</summary>

![boot-screen](https://github.com/user-attachments/assets/e406397f-af6c-4cb7-9598-b91e7b0c939c)

</details>

<details>
<summary>Partition the OS disk (this will also install the OS)</summary>

![partition-1](https://github.com/user-attachments/assets/93638267-a4e6-4e2e-a721-5dfcc31104b6)
![partition-2](https://github.com/user-attachments/assets/2e479121-90e1-4d84-8b39-64c62888d384)
![partition-3](https://github.com/user-attachments/assets/1b10c734-88ea-4bd9-9f33-7aa57abd0583)
![partition-4](https://github.com/user-attachments/assets/8ad5f872-cf9e-454b-b34c-ea303aa080e0)
![partition-5](https://github.com/user-attachments/assets/04c9b975-534f-4d3d-aa5d-05cf5ed65c5e)

</details>

</details>

<details>
<summary>Install GRUB boot loader on OS disk</summary>

![grub-1](https://github.com/user-attachments/assets/09868bf7-02dc-4cbd-aa1c-5ff3d8a9c03e)

</details>

<details>
<summary>Stop the VM when installation is complete</summary>

![hard-stop](https://github.com/user-attachments/assets/abb6e192-d4be-4a6e-a2cc-e30a3a2bb357)

</details>

<details>
<summary>Change VM boot order to boot from OS disk first</summary>

![boot-order](https://github.com/user-attachments/assets/0bd42f22-1a13-4902-b75b-5361f175386a)

</details>

<details>
<summary>Start the VM and follow setup steps. <b>Important: Domain must be set to the IP address of the VM</b></summary>

![turnkey-1](https://github.com/user-attachments/assets/8b7c3a5f-3dff-4225-a043-eb5a378d7a0e)
![turnkey-2](https://github.com/user-attachments/assets/ae29c026-48b2-4872-a6eb-0c02df526a32)
![turnkey-3](https://github.com/user-attachments/assets/6377d458-1cb3-4180-80bc-2bd4e9818e28)

</details>

### Your Turnkey Nextcloud VM is up and running, congrats!

![turnkey-4](https://github.com/user-attachments/assets/ad42d5f1-e114-4aff-a092-90f843af7550)

**Comment by Bob-Tek** [comment]
CAn anyone mention how this is better than the container script?

**Reply by pfeiffr** [reply]
Overall higher security, especially if exposed, by having higher isolation from the host system compared to the LXC container.

**Reply by SloppyEngineer** [reply]
> CAn anyone mention how this is better than the container script?

Also additional compatibility, for example mounting CIFS/NFS shares without having to map IDs or change ownership

**Comment by Levtos-Acc** [comment]
Hello @havardthom,

First of all, thank you for providing the Proxmox VE Nextcloud VM script—it’s been very helpful! However, I’ve encountered an issue during the setup process that prevents me from using the server after installation.

It seems that the script defaults to a US keyboard layout during the initial setup. After installation, the keyboard layout switches to the system’s default (e.g., German in my case). This causes problems when special characters like @, =, or \ are used in the password during configuration. These characters are mapped differently in other layouts, and as a result, the password becomes unusable post-installation, making it impossible to log in to the server.

Here are the details of my setup:

    Environment: TurnKey Linux (from your provided script)
    Keyboard layout: German (system default post-installation)
    Problem: Password mismatch due to the initial setup defaulting to a US keyboard layout.

To resolve this issue, I suggest one of the following solutions:

    Add a prompt during the installation process to allow users to select their preferred keyboard layout before entering sensitive data like passwords.
    Set a default password (e.g., admin or changeme) during installation, with instructions to change it after the setup is complete.

Either of these changes would ensure users can log in to the server after installation without running into this issue.

Thank you for your time and support! Please let me know if you need any further details to address this.

Best regards,

**Reply by hdortim** [reply]
This works for me:
- Change the keyboard layout to US temporarly
- enter the password in an editor with the german keys on the changed keyboard layout
- note down the password with the changed characters
- and try it to login

**Reply by hyzteric** [reply]
What I did is go to wikipedia and look at a qwerty keyboard layout and compared it to my belgian azerty keyboard so I could use the same keys.
I used Ertyuiojk8
What went wrong is that as for a number in the password I chose the number 8 ... but with the current keyboard layout during setup it was actually an * that got in... so my password that needed at least one lower case, one uppercase and one number happened to not have any number in.
So I could log in with user admin and my password Ertyuiojk* instead of Ertyuiojk8

**Comment by FriedrichVanWel** [comment]
The installation is failing when trying to install GRUB.

- proxmox-ve: 8.2.0
- 4 x Intel(R) Core(TM) i5-6500T CPU @ 2.50GHz (1 Socket)
- Linux 6.8.8-2-pve (2024-06-24T09:00Z)

**Reply by havardthom** [reply]
Make sure you are targeting the OS disk which is partitioned, it will fail if it tries to install GRUB on the ISO image disk. On `Install GRUB boot loader on OS disk` step you can also select `no` option and then manually select which disk, e.g. `sda` or `sdb` depending on which is your partitioned OS disk.

**Reply by FriedrichVanWel** [reply]
Thank you havardthom. I was able to pick the correct disk, and everything worked fine from there.

**Reply by eagle275** [reply]
That is a dumb screen ... 
when I select "no" .. I expect to stop installation of grub.. not give me manual control.. instead the stupid installer should "know" that the drive it just installed every other component onto is the one for the grub installer as well .. 
I tried to manually do it from the "live cd" but then i get more errors .. 

And even after manually getting grub installed correctly (at least according to the installer) .. it doesnt work 
>Booting from Hard Disk...
>Grub loading...
>Welcome to Grub!

>error: attempt to read or write outside of disk 'hd0' .

What do I enter now?

**Reply by Sideflare** [reply]
installing grub to the correct 12gb disk failed for me

**Reply by Sideflare** [reply]
addind a few gbs to the larger disk and selecting 90% default fixed it.   grub installed...

**Reply by thisisjaymehta** [reply]
Mine keep going to Grub Rescue

**Reply by tserra30** [reply]
I got the same thing installing grub on the correct disk that is in the gb not mb

**Comment by Praaug** [comment]
Hey @havardthom ,

thank you for the detailed installation instrcutions.
I have the problem, that I can't login with the `admin` or `adminer` users in nextcloud itself, the webmin or the db login. The root login for the terminal is working.

I have reinstalled mutiple times, also with setting up very easy passwords, that are not subject to the different keyboard layouts like `1234Something` but I always get the error message, that username or password are wrong.

I also tried to reset the admin password through the terminal, but there I get the message that occ is only allowed with the user that created it, www-data. `sudo` and `su` are not working as `sudo` is not installed.

Does anyone have any idea, where I did something wrong, or what I can do to be able to login and finish the setup?

**Reply by markrgriffin** [reply]
I am in the same situation, install ok and can log in as root to the terminal, however I cannot get past the webpage login or webmin or Adminer. Anyone got a solution for this?

**Reply by kpatient** [reply]
I had this issue and was due to shift key not being passed through VM. So every password character I thought was uppercase was actually lowercase.

As others have mentioned there might also be other subtle differences either due to the img's default keyboard layout and how keyboard inputs are passed through the console.

**Reply by martin8310** [reply]
Really frustrating, the VM is running and it's failing because of that stupid login, what a shame. Unfortunately, this script is useless to me. But thanks for trying.

**Reply by MinionCZ** [reply]
Hello,
had the same issue as you did with getting unathorized logging into webmin interface. Helped me to log into console in VM as root and type this command, which changes password for user root for webmin interface.
```bash
/usr/share/webmin/changepass.pl /etc/webmin root NEWPASSWORDHERE
```

I found the solution [here](https://forum.virtualmin.com/t/webmin-login-suddenly-fails/129069/3) if you want to take a look.

**Comment by iukbox** [comment]
Using Advanced Settings
Virtual Machine ID: 101
Using Machine Type: i440fx
Using Disk Cache: None
Using Hostname: turnkey-nextcloud-vm
Using CPU Model: KVM64
Allocated Cores: 2
Allocated RAM: 2048
Using Bridge: vmbr0
Using MAC Address: 02:CD:45:8E:EA:41
Using Vlan: Default
Using Interface MTU Size: Default
Start VM when completed: yes
Creating a TurnKey Nexcloud VM using the above advanced settings
 ✓ Using local-lvm for Storage Location.
 ✓ Virtual Machine ID is 101.
 ✓ http://mirror.turnkeylinux.org/turnkeylinux/images/iso/turnkey-nextcloud-18.0-bookworm-amd64.iso
 ✓ Downloaded turnkey-nextcloud-18.0-bookworm-amd64.iso
 - Creating a TurnKey Nexcloud VM...
**_[ERROR] in line 405: exit code 5: while executing command pvesm alloc $STORAGE $VMID $DISK0 4M >&/dev/null_**

**Reply by kotovasia5120** [reply]
I didn't have the numbers of containers and disks for them. I manually assigned another container number with an unoccupied disk number.

**Comment by zientas** [comment]
I installed it and everything started up nicely, but I have questions:
1. how to add face recognition models, occ commands don't work
2. is it possible to copy photos from the shell level to nestcloud directories and have it recognize them? I have a few thousand photos and adding them via the browser is tiring. When I copied a few photos from the shell level, nextcloud doesn't see them.

**Reply by Gorkovets** [reply]
> Я установил его, и все заработало хорошо, но у меня есть вопросы:
> 
> 1. как добавить модели распознавания лиц, команды occ не работают
> 2. можно ли копировать фотографии с уровня оболочки в каталоги nestcloud и распознавать их? У меня несколько тысяч фотографий, и добавлять их через браузер утомительно. Когда я скопировал несколько фотографий с уровня оболочки, nextcloud их не видит.

buddy Use this command  =>   turnkey-occ recognize:download-models

**Comment by MikeAnita** [comment]
Hello,

I installed the VM successfully and got to the login page.
What are the login details here?
![Screenshot 2025-01-12 112358](https://github.com/user-attachments/assets/d68352b3-aba6-44f4-a3ef-0f4188aaef0d)

**Reply by renefichtmueller** [reply]
same here, installed the LXC with Alpine but have Probs with the Login

**Reply by Gorkovets** [reply]
Only Alpine: To get the username and password, run the script again inside the LXC shell.

**Reply by Node815** [reply]
For the VM, use this command and it will allow you to reset the password for the admin account:

```sudo -u www-data php /var/www/nextcloud/occ user:resetpassword admin```

**NOTE:** you most likely won't have sudo installed so you will need to run that.  I ran into this error upon a fresh install.

**Reply by dizir71** [reply]
hello, 
i told me sudo not installed, how to run the script to reset the admin password, i have a fresh install, can access the admin in the shell but no access to the admin on the website, nor can i access the database, first i thought it was the keyboard (german) but the i did not use any other characters then english and number, still access denied. can you tell me what is wrong or how can i run that script ? with runuser maybe? thank you or you can tell me to run "sudo command" or install it?

**Reply by S-lax** [reply]
Hi,
I also had the issue. Turns out the passwords were set (by me) in english keyboard layout.

You can login with the set nextcloud admin pw.

**Reply by ttrueten** [reply]
You can use the occ App to run commands.

> hello, i told me sudo not installed, how to run the script to reset the admin password, i have a fresh install, can access the admin in the shell but no access to the admin on the website, nor can i access the database, first i thought it was the keyboard (german) but the i did not use any other characters then english and number, still access denied. can you tell me what is wrong or how can i run that script ? with runuser maybe? thank you or you can tell me to run "sudo command" or install it?

**Comment by JRLEMOYNE** [comment]
Hi, thanks for the easy setup guide. Can I add my NFS share from my NAS to the VM so it stores the files there? Tried to find some informations on how to do it but didn't find anything helpful, so if anyone can help me that would be greatly appreciated!

**Reply by renefichtmueller** [reply]
Which NAS you are using? I've added my Synology NAS as a storage system

**Reply by JRLEMOYNE** [reply]
It's custom running truenas scale. I already have my shares linked as storage in proxmox and use them with some lxc but I don't know if I can add them to the nextcloud vm the same way.

**Reply by nanohits** [reply]
Same here I have TrueNAS scale but not sure how to add that as a storage device?

**Reply by SpacemannFinn** [reply]
Did you ever figure out how to do this?

**Reply by JRLEMOYNE** [reply]
No, didn't have much time to search for answers. Still hope someone can help. I'll make sure to post the solution here if I find it.

**Reply by SpacemannFinn** [reply]
Aww that sucks. I am trying to solve this issue with a ZFS pool I made in proxmox

**Reply by wtangofoxtrot** [reply]
I think the solution you want is:

1. Make your nextcloud dataset in truenas accessible via a nfs share
2. In the VM mount that NFS share locally over the directory where nextcloud stores its data and then add it to `/etc/fstab` so it retains the share on reboot.

**Reply by kevincmach** [reply]
Is anyone still interested in this? I got it working and i wrote a guide. For reference I iuse Proxmox 8 and TrueNAS Scale as my backing NAS

1) Create TrueNAS dataset
2) Enable NFS share on dataset
3) Install NFS client in VM
4) Create mount point in VM
5) Mount NFS share manually
6)Add NFS mount to /etc/fstab
7) Bind-mount to Nextcloud’s
8) Set Permissions 
9) Update Nextcloud config.php
10) Reboot

drop me a DM if anyone needs help with this setup

**Reply by ryohiroko** [reply]
@kevincmach would u like to help to give detail from step 6 until 9 ? i already step add nfs in etc/fstab

**Reply by remcomeeder** [reply]
@kevincmach I am interested to see how you have solved it. I am in the process of setting up basically the same. I have a Proxmox server which hosts my TrueNAS instance.

**Reply by alexgarcia19** [reply]
@kevincmach can you send me the guide? TIA

**Comment by emircanerkul** [comment]
Turnkey only allow me to download at 50kbps. Also it is version 18. What year they living on? Currently v31 released

**Comment by WH4MGMT** [comment]
can you remove the extra hardisk used for install?

**Reply by LeChat109** [reply]
yes you can, as soon as you finished the installation process

**Comment by dualizeo** [comment]
Wouldn't the AIO Edition be more efficient AND better? ( All in One ) Even NextCloud officially uses that https://github.com/nextcloud/all-in-one

**Reply by davemenkehorst** [reply]
I miss Borg backup and other solutions. Is the All in One version better?

**Comment by jolypas** [comment]
Hello
Any way to change keyboard layout during installation (complicate for all passwords setup with wrong keyboard) ?

**Comment by steve41200** [comment]
Anyone who how join this VM via reverse proxy ? (NginxProxy Manager) ?
I joint all of my virtual machine except here (too many requests) if I join 80 in Http or 443 in https

**Reply by writzthecat** [reply]
Having the some Problem. Even if i change for the thrusted proxies it's not going to work :;(

**Reply by Wajalot** [reply]
> Anyone who how join this VM via reverse proxy ? (NginxProxy Manager) ? I joint all of my virtual machine except here (too many requests) if I join 80 in Http or 443 in https

Maybe is late, but to others. This worked for me.

Assuming ports 80 and 443 are targeting to the NginX Proxy Manager, I modified both /etc/apache2/ports.conf and /etc/apache2/sites-enabled/nextcloud.conf changing 8080 and 8443 instead 80 and 443 respectively, then I restarted apache2 and set those ports in NginX Proxy Manager. Actually just the 8443, with Lets encrypt certificated generated and linked and forcing https. Also the host with https.

**Reply by s-a-kuehl** [reply]
Thank you Wajalot, this worked for me as well and saved me a lot of trouble!

**Comment by pvd276** [comment]
not working with hardisk over 2TB

**Reply by Wajalot** [reply]
I installed Nextcloud following the guide, however I added a disk with more than 2TB mounted in /var/... for data storing.

**Comment by ryosoftware** [comment]
Hello,
What are the beneficts from this VM over a unprivileged LXC container then install NC via a docker inside?
Thanks

**Reply by Wajalot** [reply]
It never worked finely for me as LXC. However with VM pretty fine, settings and store. Also webdav and other services works much better with VM, at least in my experience.

**Comment by vitalikmorozz** [comment]
After following the instructions here, it all worked fine when I was trying to access it through the IP of the VM. I'm using Cloudflare tunnel (cloudflared) for accessing my locally hosted things, and after setting up a subdomain and as a target my VM's IP (HTTP), when trying to access it through the domain if went into the redirect loop. I tried targetting HTTPS, but it doesn't work at all. So for anyone who's also struggling to make it work, here's how I solved it:
- from the file /etc/apache2/sites-enabled/nextcloud.conf remove/comment the following lines
> RewriteCond %{HTTPS} !=on
> RewriteRule ^ https://%{SERVER_NAME}%{REQUEST_URI} [L,R=301]

- restart apache2 service
> systemctl restart apache2

- in the `/var/www/nextcloud/config/config.php` add your domain to `trusted_domains`, `overwrite.cli.url`, `overwritehost`, set `overwriteprotocol` to https. Use [this ](https://docs.nextcloud.com/server/latest/admin_manual/configuration_server/reverse_proxy_configuration.html)guide for reference.

- in cloudflare tunnel config pannel, point the domain to HTTP IP of your VM.

I'm new to self hosting and maybe did something wrong, but it seems to work. http -> https redirects still work, I guess that is handled on the cloudflare side. I honestly have no experience with apache, so if there's a better way to do it, let me know.

**Comment by DeonXL** [comment]
Hi, can you tell me how to connect a shared folder via SMB? For some reason, it says the password is incorrect, even though I'm entering everything correctly.

**Reply by ghost** [reply]
Maybe this could help: https://forum.proxmox.com/threads/tutorial-unprivileged-lxcs-mount-cifs-shares.101795/

**Reply by Bazz-Man** [reply]
Unfortunately this is a known bug and has been around for a while, such a shame for what seems a simple function.  Think the way forward is to mount locally vi fstab and then use the local share function in nextcloud
see for more info https://github.com/nextcloud/server/issues/53472

**Reply by DeonXL** [reply]
Мне помогло установить на сервер php-smbclient

**Comment by DeNohvo** [comment]
Whats the best way to install this so data is stored on a separate disk? All my VM/LXC are on a small SSD and data is on a zfs pool. 

So would like boot VM OS from my SSD but keep on data on zfs.

**Reply by Wajalot** [reply]
Maybe bit late, sorry, but for others. In my case I have a ZFS in Proxmox. So, when I created the VM, I stopped before installation of Turnkey and then I created a SATA disk with part of that ZFS pool. In picture below this lines, sata0 disk, Almacen (store).

<img width="633" height="474" alt="ncsata" src="https://github.com/user-attachments/assets/bcdc8e01-708a-435e-a736-ad379ea3b2dd" />

So, during installation of Turnkey, system root was set in scsi0, local-VM,  however I created a partition volume mounting the directory for Nextcloud data, /var/www/nextcloud-data, in that SATA disk.

**Comment by Aholicknight** [comment]
![nextcloudwgetdownload](https://github.com/user-attachments/assets/2fd9f79b-88fb-4a2f-9889-a820902cf9b3)

Is anyone else getting the same problem when trying to upgrade after installing? On the website updater it is failing and using wget also fails.

**Reply by Eslam-Moustafa-Ramadan** [reply]
The same happens with me

**Reply by Aholicknight** [reply]
@Eslam-Moustafa-Ramadan To fix this issue I had to cancel the web updater over and over till it finally downloaded. Took me 5 tries.

**Comment by n3de** [comment]
The installation works fine and I can login to nextcloud and to the cli, but I can't login to the trunkey administration via ipv4:12321 I check my keyboard layout and diffrent user (root, admin, adminer) but nothing helpfs. What do I wrong and how to solve the problem?

**Comment by TheCustomFHD** [comment]
This currently gets an F rating on https://scan.nextcloud.com due to being VERY outdated. v29. Also the "updateable" flag seems wrong, as i couldn't find an easy update path, update command doesnt exist, and the script ran inside the VM doesnt make sense.

Is this script abandoned?

**Reply by MickLesk** [reply]
https://docs.nextcloud.com/server/32/admin_manual/maintenance/upgrade.html

**Reply by MickLesk** [reply]
Turnkey is not from us, its delivered by an third party proxmox.

**Reply by TheCustomFHD** [reply]
I did once try to update the old Nextcloud instance and it kinda just complained about a bad hash in one of the core files.

I also cant Imagine that the Turnkey images are that old, but possible, i didnt check. 

There is the nextcloud/vm install production script though, im surprised that that isnt used here.

**Reply by Wajalot** [reply]
I used it once again last week and it worked fine for me. It is true Nextcloud is very outdated in this script, however just running upgrade from administration screen it does very fast. 29 to 30, then 30 to 31, and so on until last  version, 33. Also all flags were fixed. You can follow this guide for it, I did it and worked finely for me.

https://wiki.hakedev.com/docs/proxmox/nextcloud-post-install/

**Reply by hayvan96** [reply]
Hi, I've upgraded Nextcloud up to 32 so far, but I see 33 requires php8.5 (8.2 is deprecated but supported). As turnkey18 is based on bookworm, how are we supposed to upgrade php? By a onplace Debian upgrade?

**Reply by MickLesk** [reply]
With apt Sources. Use the official PHP source and Upgrade it

**Reply by hayvan96** [reply]
Well, for PHP, NC33 still works with 8.2, but I guess when we upgrade to 8.5, both versions will sit next to each other, so updating apt sources+apt upgrade won't be enough (telling apache, for instance, which PHP to use). Has anybody tried to upgrade PHP via their own sources, or an inplace debian (-->trixie) upgrade?

**Comment by sashadeg** [comment]
After upgrading from Nextcloud 30.0.17 to 31.x, login no longer works.

Symptoms:
Login page loads normally
Username and password are accepted
After submitting credentials:
page reloads
user is returned to login screen
No visible error message is shown
No 2FA prompt appears
Issue affects all browsers and devices
Incognito/private mode does not help

Reproducibility
The issue is 100% reproducible:
Start with Nextcloud 30.0.17 → login works normally
Upgrade to 31.x
Attempt login
Result: login loop (silent redirect back to login page)

Expected behavior
User should be authenticated and redirected to the dashboard after successful login.

Actual behavior

Authentication appears successful, but session is not established and user is redirected back to the login page without any error.

Things already checked
Redis is working correctly (PING OK)
PHP Redis extension is installed and active
Memcache configured with Redis (local + locking)
All apps disabled except core ones → issue persists
2FA (twofactor_totp) disabled → no effect
Issue is not browser-related
Downgrading back to 30.0.17 immediately fixes the issue

**Comment by ImTheSpyke** [comment]
During VM install i hit the error "Unable to install GRUB".
The fix is to use in VM settings Machine "i440fx" instead of "q35".

You're welcome !

**Comment by ha11oga11o** [comment]
Hello,

it seems Turnkey Linux appliance got new release. Can it be integrated to script? This goes with turnkey-nextcloud-18.1-bookworm-amd64.iso. It seems new is released; turnkey-nextcloud-19.0-trixie-amd64.iso

https://www.turnkeylinux.org/nextcloud/

Thank you.

**Comment by ernie95** [comment]
Hi All

Does the script now base on Debian 13 ?
https://github.com/community-scripts/ProxmoxVE/blob/main/vm/nextcloud-vm.sh

If not, will be there an update ?

Thanks

**Reply by MickLesk** [reply]
In next weeks, you can replace ProxmoxVE to ProxmoxVED in the bash call, thats the Dev Script. Feedback Welcome.
